test/integration/http/websockets-test.js

```diff
@@ -1,68 +1,100 @@
-    teardown(t)
-  })
-  ws.close()
+let sendMessage = payload => {
+  _ws.send(JSON.stringify(payload))
 }
 
-function connect (t, payload) {
-  t.pass('WebSocket connected')
-  ws.send(JSON.stringify(payload))
+let nextEvent = async t => {
+  let request = await _events.nextRequest()
+  t.ok(true, 'got next request')
+  return request
 }
 
-function message (t, m) {
-  let message = JSON.parse(m)
-  t.ok(message, 'Received message payload from WS Lambda confirming send')
-  console.log(`WebSocket message:`, message)
+let startupAndConnect = async () => {
+  _events.reset()
+  await connectWebSocket()
+}
 
-  let { event, req } = message
-  t.ok(req.requestContext.connectionId, 'Got request context with connectionId')
+let expectFunctionBasics = (t, websocketMessage, expectedFunctionName) => {
+  let { functionName, event } = websocketMessage
+  t.equal(functionName, expectedFunctionName, `[${expectedFunctionName}] Invoked correct function`)
+  t.equal(event.requestContext.routeKey, `$${expectedFunctionName}`, `[${expectedFunctionName}] Included correct routeKey`)
+  t.equal(event.isBase64Encoded, false, `[${expectedFunctionName}] isBase64Encoded is false`)
+}
 
-  let body = JSON.parse(req.body)
-  t.equal(event, body.expect, `Invoked correct WebSocket Lambda / action: ${event}`)
-  t.equal(body.hi, data.hi, 'Got body payload')
-  t.equal(req.isBase64Encoded, false, 'Got isBase64Encoded false')
+let expectPayload = (t, websocketMessage, expectedBody) => {
+  let { functionName, event: { body } } = websocketMessage
+  if (body === undefined || expectedBody === undefined) {
+    t.equal(body, expectedBody, `[${functionName}] Expected undefined body`)
+    return
+  }
+
+  let parsedBody = JSON.parse(body)
+  if (typeof parsedBody !== 'object') {
+    t.equal(parsedBody, expectedBody, `[${functionName}] Expected non object body matched lambda input`)
+    return
+  }
 
-  close(t)
+  // shallow check event object which is fine for our purposes and will throw if we nest objects
+  for (let [ key, value ] of Object.entries(expectedBody)) {
+    t.deepEquals(parsedBody[key], value, `[${functionName}] Expected body.${key} to match lambda input`)
+  }
 }
 
+let expectRequestContext = (t, websocketMessage, expectedContext = {}) => {
+  let { functionName, event: { requestContext } } = websocketMessage
+  t.ok(requestContext.connectionId, `[${functionName}]  Got requestContext with connectionId`)
+  t.ok(requestContext.requestId, `[${functionName}] Got requestContext with requestId`)
+  t.equal(requestContext.messageDirection, 'IN', `[${functionName}] Got requestContext correct messageDirection`)
 
-test('Set up env', t => {
+  for (let [ key, value ] of Object.entries(expectedContext)) {
+    t.deepEquals(requestContext[key], value, `[${functionName}] Expected requestContext.${key} to match expected input`)
+  }
+}
+
+test('Set up env', async t => {
   t.plan(1)
-  t.ok(sandbox, 'Got sandbox')
+  _events = await makeSideChannel()
```

Comment:
For consistency / catching meta boogs, suggest also adding `t.ok(_events, ...`

Reply:
Technically we can use `t.end()` but using `t.plan()` is super deliberate. It ensures a completely quantified and deterministic test run – no extra tests can be run or missed. Fwiw!